File: frontend/src/types/payroll.ts
// AI-HRMS — Payroll module TypeScript types

// ─── Enums ────────────────────────────────────────────────────────────────────

export type PayrollRunStatus =
  | 'draft'
  | 'processing'
  | 'processed'
  | 'approved'
  | 'paid'
  | 'cancelled'
  | 'rejected';

export type PayrollRecordStatus = 'pending' | 'processed' | 'paid' | 'on_hold';

export type ApprovalAction = 'approve' | 'reject';

// ─── Tax Slab ─────────────────────────────────────────────────────────────────

export interface TaxSlab {
  id:          string;
  year:        number;
  min_income:  number;
  max_income:  number | null;
  tax_rate:    number;  // 0.0 – 1.0
  fixed_tax:   number;
  description: string | null;
  is_active:   boolean;
  created_at:  string;
  updated_at:  string;
}

export interface TaxSlabCreate {
  year:        number;
  min_income:  number;
  max_income:  number | null;
  tax_rate:    number;
  fixed_tax:   number;
  description: string | null;
  is_active:   boolean;
}

export type TaxSlabUpdate = Partial<Omit<TaxSlabCreate, 'year'>>;

// ─── Employee Minimal ─────────────────────────────────────────────────────────

export interface EmployeeMinimal {
  id:                string;
  employee_code:     string;
  full_name:         string;
  department_name:   string | null;
  designation_title: string | null;
  photo_url:         string | null;
}

// ─── Payroll Run ──────────────────────────────────────────────────────────────

export interface PayrollRun {
  id:                    string;
  month:                 number;
  year:                  number;
  label:                 string | null;
  status:                PayrollRunStatus;
  total_employees:       number;
  total_gross:           number;
  total_net:             number;
  total_deductions:      number;
  total_eobi_employee:   number;
  total_eobi_employer:   number;
  total_income_tax:      number;
  processed_by:          string | null;
  approved_by:           string | null;
  run_at:                string | null;
  approved_at:           string | null;
  paid_at:               string | null;
  notes:                 string | null;
  celery_task_id:        string | null;
  created_at:            string;
  updated_at:            string;
}

export interface PayrollRunCreate {
  month:          number;
  year:           number;
  department_ids: string[] | null;
  notes:          string | null;
}

export interface PayrollRunListResponse {
  count:   number;
  results: PayrollRun[];
}

// ─── Payroll Record ───────────────────────────────────────────────────────────

export interface PayrollRecord {
  id:                   string;
  payroll_run_id:       string;
  employee_id:          string;
  employee:             EmployeeMinimal;

  // Earnings
  basic_salary:         number;
  house_rent_allowance: number;
  medical_allowance:    number;
  transport_allowance:  number;
  fuel_allowance:       number;
  other_allowances:     Record<string, number> | null;
  total_allowances:     number;
  gross_salary:         number;

  // Deductions
  eobi_employee:        number;
  eobi_employer:        number;
  sessi:                number;
  income_tax:           number;
  loan_deduction:       number;
  advance_deduction:    number;
  other_deductions:     Record<string, number> | null;
  total_deductions:     number;
  net_salary:           number;

  // Attendance
  working_days:         number;
  present_days:         number;
  absent_days:          number;
  late_days:            number;
  overtime_hours:       number | null;
  paid_leave_days:      number;
  unpaid_leave_days:    number;
  is_prorated:          boolean;

  payslip_url:          string | null;
  status:               PayrollRecordStatus;
  created_at:           string;
  updated_at:           string;
}

export interface PayrollRunDetail extends PayrollRun {
  records: PayrollRecord[];
}

// ─── Payslip Data ─────────────────────────────────────────────────────────────

export interface AllowanceItem {
  name:   string;
  amount: number;
}

export interface DeductionItem {
  name:   string;
  amount: number;
}

export interface PayslipData {
  record_id:       string;
  payroll_run_id:  string;
  month:           number;
  year:            number;
  employee_id:     string;
  employee_code:   string;
  full_name:       string;
  designation:     string | null;
  department:      string | null;
  cnic:            string | null;
  joining_date:    string | null;
  bank_name:       string | null;
  account_number:  string | null;
  iban:            string | null;
  basic_salary:    number;
  allowances:      AllowanceItem[];
  total_allowances: number;
  gross_salary:    number;
  deductions:      DeductionItem[];
  total_deductions: number;
  net_salary:      number;
  working_days:    number;
  present_days:    number;
  absent_days:     number;
  paid_leave_days: number;
  overtime_hours:  number | null;
  overtime_amount: number;
  generated_at:    string;
  payslip_url:     string | null;
}

// ─── Salary Preview ───────────────────────────────────────────────────────────

export interface SalaryBreakdown {
  employee_id:            string;
  employee_name:          string;
  month:                  number;
  year:                   number;
  basic_salary:           number;
  total_allowances:       number;
  gross_salary:           number;
  eobi_employee:          number;
  eobi_employer:          number;
  income_tax:             number;
  loan_deduction:         number;
  advance_deduction:      number;
  other_deductions_total: number;
  total_deductions:       number;
  net_salary:             number;
  working_days:           number;
  present_days:           number;
  absent_days:            number;
  overtime_hours:         number;
  overtime_amount:        number;
  effective_tax_rate:     number;
  annual_gross:           number;
  annual_tax:             number;
}

// ─── Bank File ────────────────────────────────────────────────────────────────

export interface BankFileEntry {
  employee_code:  string;
  employee_name:  string;
  bank_name:      string;
  account_number: string;
  iban:           string | null;
  net_salary:     number;
  reference:      string;
}

// ─── Filters ──────────────────────────────────────────────────────────────────

export interface PayrollFilterParams {
  month?:        number;
  year?:         number;
  department_id?: string;
  status?:       PayrollRunStatus;
  employee_id?:  string;
  page?:         number;
  page_size?:    number;
}

// ─── Approval ─────────────────────────────────────────────────────────────────

export interface PayrollApprovalRequest {
  action: ApprovalAction;
  notes?: string | null;
}

File: backend/app/models/payroll.py
"""
AI-HRMS — PayrollRun, PayrollRecord, TaxSlab models
"""

import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, Enum, ForeignKey, Integer, Numeric, String, Text, text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin, UUIDPrimaryKeyMixin

if TYPE_CHECKING:
    from app.models.employee import Employee
    from app.models.tenant import Tenant, User


# ─── ENUMs ────────────────────────────────────────────────────────────────────

PayrollRunStatusEnum = Enum(
    "draft", "processing", "approved", "paid", "cancelled",
    name="payroll_run_status_enum",
)

PayrollRecordStatusEnum = Enum(
    "pending", "processed", "paid", "on_hold",
    name="payroll_record_status_enum",
)


# ─── Payroll Run ──────────────────────────────────────────────────────────────

class PayrollRun(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    """
    A monthly payroll batch. One run per (tenant, month, year).
    All individual PayrollRecords are linked to a run.
    """
    __tablename__ = "payroll_runs"

    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("tenants.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    month: Mapped[int] = mapped_column(Integer, nullable=False)   # 1–12
    year: Mapped[int] = mapped_column(Integer, nullable=False)
    label: Mapped[str | None] = mapped_column(String(100), nullable=True)  # e.g. "March 2026 Payroll"

    status: Mapped[str] = mapped_column(
        PayrollRunStatusEnum, nullable=False, server_default="draft", index=True,
    )

    # Aggregates (computed and stored for fast reporting)
    total_employees: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    total_gross: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    total_net: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    total_deductions: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    total_eobi_employee: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    total_eobi_employer: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    total_income_tax: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)

    # Workflow
    processed_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True,
    )
    approved_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True,
    )
    run_at: Mapped[datetime | None] = mapped_column(nullable=True)
    approved_at: Mapped[datetime | None] = mapped_column(nullable=True)
    paid_at: Mapped[datetime | None] = mapped_column(nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Celery task ID for async processing
    celery_task_id: Mapped[str | None] = mapped_column(String(255), nullable=True)

    # Relationships
    tenant: Mapped["Tenant"] = relationship("Tenant", lazy="noload")
    records: Mapped[list["PayrollRecord"]] = relationship(
        "PayrollRecord", back_populates="payroll_run", cascade="all, delete-orphan",
    )
    processed_by_user: Mapped["User | None"] = relationship(
        "User", foreign_keys=[processed_by], lazy="noload",
    )

    def __repr__(self) -> str:
        return f"<PayrollRun {self.year}-{self.month:02d} status={self.status!r}>"


# ─── Payroll Record ───────────────────────────────────────────────────────────

class PayrollRecord(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    """
    Individual payroll calculation for one employee in one payroll run.
    One record per (payroll_run_id, employee_id).
    """
    __tablename__ = "payroll_records"

    payroll_run_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("payroll_runs.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    employee_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("employees.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # ── Earnings ──────────────────────────────────────────────────────────────
    basic_salary: Mapped[int] = mapped_column(Integer, nullable=False)
    house_rent_allowance: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    medical_allowance: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    transport_allowance: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    fuel_allowance: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    other_allowances: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    total_allowances: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    gross_salary: Mapped[int] = mapped_column(Integer, nullable=False)

    # ── Deductions ────────────────────────────────────────────────────────────
    eobi_employee: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    eobi_employer: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    sessi: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    income_tax: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    loan_deduction: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    advance_deduction: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    # Flexible additional deductions (e.g. penalties)
    other_deductions: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    total_deductions: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)

    net_salary: Mapped[int] = mapped_column(Integer, nullable=False)

    # ── Attendance Summary ────────────────────────────────────────────────────
    working_days: Mapped[int] = mapped_column(Integer, nullable=False)
    present_days: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    absent_days: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    late_days: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    overtime_hours: Mapped[float | None] = mapped_column(Numeric(6, 2), nullable=True)

    # Leave summary for this pay period
    paid_leave_days: Mapped[float] = mapped_column(Numeric(4, 1), server_default=text("0"), nullable=False)
    unpaid_leave_days: Mapped[float] = mapped_column(Numeric(4, 1), server_default=text("0"), nullable=False)

    # Prorated salary flag (for joiners/leavers mid-month)
    is_prorated: Mapped[bool] = mapped_column(Boolean, server_default=text("false"), nullable=False)

    # ── Output ────────────────────────────────────────────────────────────────
    payslip_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    status: Mapped[str] = mapped_column(
        PayrollRecordStatusEnum, nullable=False, server_default="pending", index=True,
    )

    # Relationships
    payroll_run: Mapped["PayrollRun"] = relationship("PayrollRun", back_populates="records")
    employee: Mapped["Employee"] = relationship("Employee", back_populates="payroll_records")

    def __repr__(self) -> str:
        return (
            f"<PayrollRecord run_id={self.payroll_run_id!r} "
            f"employee_id={self.employee_id!r} net={self.net_salary!r}>"
        )


# ─── Tax Slab ─────────────────────────────────────────────────────────────────

class TaxSlab(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    """
    Income tax slabs for a given fiscal year (Pakistan FBR).
    Stored in DB so HR can update them without a code deploy.
    """
    __tablename__ = "tax_slabs"

    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("tenants.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    year: Mapped[int] = mapped_column(Integer, nullable=False, index=True)  # Fiscal year e.g. 2026
    min_income: Mapped[int] = mapped_column(Integer, nullable=False)         # Annual income lower bound
    max_income: Mapped[int | None] = mapped_column(Integer, nullable=True)   # NULL = no upper limit
    tax_rate: Mapped[float] = mapped_column(Numeric(5, 4), nullable=False)   # e.g. 0.2500 = 25%
    fixed_tax: Mapped[int] = mapped_column(Integer, server_default=text("0"), nullable=False)
    description: Mapped[str | None] = mapped_column(String(200), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, server_default=text("true"), nullable=False)

    def __repr__(self) -> str:
        return (
            f"<TaxSlab year={self.year!r} "
            f"range={self.min_income}–{self.max_income or '∞'} rate={self.tax_rate!r}>"
        )
